### ISSUE TITLE
Document the meaning of deptypes

### ISSUE THREAD

**Issue by citibeth** [spack/spack#3768]
@tgamblin @krafczyk @davydden @certik @scheibelp 

In #3347 and similar, I'm coming to believe we don't really know what we mean by build, link and run dependencies.  I'm hoping we can clarify that in this thread.

First of all, I believe we need a list of the different kinds of things that might happen when A->B.  Here is a probably incomplete list:
 1. B's module loaded by A's module?
 2. B added to A's build environment?
 3. B added to A's RPATH?
 4. B built as a prerequisite for A?
 5. ...any other effects that MIGHT happen?

Next, we need a specification of how a deptype affects the above behaviours (or _should_ affect the behaviors)  For example:

deptype=build: (4), (2)
deptype=link: (3)
deptype=run: (1)

Finally, we need to figure out what, if anything, happens transitively.  If A->B->C, deptype(A->B)=x and deptype(B->C)=y, then how does that affect A's behavior?  There are only 3 deptypes, but now 9 combinations of 2-level transitive deptypes we have to decide upon.  Ideally, we won't need any special rules for transitive deptypes; but if we must, we must.

If we can get these right, then we have a definition of deptypes that we can rely upon and work with in cases like #3347.

**Comment by citibeth** [comment]
@adamjstewart

**Comment by scheibelp** [comment]
I'll take a crack at creating an initial list, I'll refine this over the next couple days. If you can edit this feel free to do so. If you can't or don't want to, feel free to leave a comment and I can update this. If someone comes up with something more comprehensive I'm also happy to make suggestions for that.

For X->Y

The following occurs for all deptypes

* Y is built as a prerequisite for X

The following may occur depending on the deptype

1. Y must be concretized with X
2. X RPATHs Y
3. X adds -I for Y's include directories when building
4. Y modifies environment for X at run time (e.g. PATH)
5. Y modifies environment for X at build time
6. Y adds to PKG_CONFIG_PATH at build time

The deptypes should have the following effects (note this is not necessarily what Spack does right now)

* include dependency: 1, 3
* link dependency: 1, 2, 3, 6
* run dependency: 4
* build dependency: 5, 6

For X->Y->Z the following deptype pairs are important (and this explains the relationship between X and Z):

* link, link: 1, 2, 3, 6 (link dependencies of link dependencies are treated the same as link dependencies)
* build, run: 5 (e.g. transitive run dependencies of build dependencies should modify PATH for X at build time)
* link, include: 1, 3 ( or possibly nothing? perhaps X should list Z as an include dependency directly for Z's header files to be included for X at build time)

**Comment by adamjstewart** [comment]
Just to clarify, `include` deptypes are "Coming Soon", right?

Also, since we're on the topic, we need to consider how to implement a `test` deptype: See #1279. Here's what I'm envisioning. Let's say we have a package X that has a `test` dependency on Y. If I run:
```
$ spack install X
$ spack install --run-tests X
```
Spack should not reinstall X because the hash should not change. But if I just run:
```
$ spack install --run-tests X
```
Spack should first install Y and then install X with tests being run. As far as how it should affect the environment/compiler wrappers, I've seen a couple suggestions:

1. Add separate `test-build`, `test-link`, and `test-run` deptypes that have the same semantic meaning as the `build`, `link`, and `run` deptypes but are only built when a user asks for them through `--run-tests`

2. Allow the package to mix n match, so `test-build` is equivalent to `type=('test', 'build')`

3. Add all modifications. `test` would have the same affect as `type=('build', 'link', 'run')`, but wouldn't affect the hash.

4. Make `test` work like `build`. I'm not sure how many test dependencies actually need to be available through the compiler wrapper. Most of the ones that I've seen are just an executable that needs to be in the `PATH`.

**Comment by scheibelp** [comment]
> Just to clarify, include deptypes are "Coming Soon", right?

Yes, they are not currently in Spack

As for what the "test" deptype should imply, I'd like to treat them exactly like the build deptype, but the example of gcc in #1279 implies to me that the libraries of test deptypes need to be available while tests are run; they shouldn't be part of RPATHs though. None of 1-6 currently accommodate this (and Spack doesn't currently support it)

**Comment by adamjstewart** [comment]
Which GCC test dep do you think it needs libraries from? I guess I assumed they were all build deptypes. Even if it's not perfect, I would love to see a `build`-like `test` deptype. Then I can start adding unit tests to all of the Python packages I care about.

**Comment by scheibelp** [comment]
> Which GCC test dep do you think it needs libraries from?

I think I'm mistaken: from glancing over the dependencies you mentioned I thought TCL might but when I think more on it I think I was wrong there. I should say I wasn't basing this off of any familiarity with gcc tests.

**Comment by adamjstewart** [comment]
I think a build-like dependency should suffice for now. Any interest in working on this? I've tried taking a look but I have no idea how our deptype/concretization logic works and it seems to be in turmoil at the moment.

**Comment by scheibelp** [comment]
> Any interest in working on this?

Yes. I will self-assign #1279. However I don't plan on starting on it for at least 3-4 weeks, in part because as you mention things are a bit in flux.

**Comment by scheibelp** [comment]
> I'll refine this over the next couple days.

At this point I consider my above summary to be comprehensive. If folks feel it is missing something or doesn't explain an important relationship please mention that here - thanks!

**Comment by alalazo** [comment]
**TL;DR** I don't think we should encourage packagers to use `depends_on('libfoo', type='link')` for usual C and C++ libraries even if it seems the most correct modeling right now. They should stick to `depends_on('libfoo')` instead and we should change the default dependency type at some point. Long story below.

----

Sorry if I came so late to this party, just finished reading the discussion above. I'll try to outline below a few high-level considerations.

#### Desired state of things

I'll start with what we want to have, i.e. 5 types of dependencies, and what I think they should mean:

1. `build`: the dependency is needed at build time and does not enter the DAG hash. These are for the most part tools needed to build the current package, e.g. `cmake` or similar.
2. `include`: the dependency is needed at build time and enters the DAG hash. Any dependency that contains headers compiled during the build of the current package should have this tag.
3. `link`: the dependency is needed at link (and load) time, enters the DAG hash and the RPATH
4. `run`: the dependency is needed at run time and enters the DAG hash
5. `test`: the dependency is needed at build time if we want to run tests on the current package, does not enter the hash

These 5 types of dependencies look very sensible to me. The only addition I can think of is that `test` dependencies might be extended to `build-test` and `run-test` depending on whether we need to have them at build time (with the stage still there) or if it is fine to have them at run-time to run the tests after a package has been installed. Correct modeling of compilers (build dependency that might inject link dependencies) is out of the scope of this comment.

#### Current state of things

Now I'll go with what we have right now, i.e. all the dependency types but `include`. My wrong understanding based on the [docs](https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types)  was that `include` types were mocked as `build` + `link` types and that `build` type dependencies were responsible for adding include paths where needed. 

What really happens is that `build` dependencies have already their final meaning - and they don't influence in any way the include paths. Contrary to what is stated in the docs `link` dependencies takes care of that. Note that this behavior wasn't there since the beginning, but was instead introduced at some point after e9d4780bbc307f37e6385fc80c8fea1abaab5782 without changing the default dependency type of packages, which is still `build` + `link`.

A couple of suprising (at least to me) results stem from the current state of affairs:

1. The default dependency type - a build tool that is also used as a library at link time - is such that no package I know of meets it exactly. Previously, when `build` was injecting include paths, this default was cut on C and C++ libraries.
2. C++ header only libraries fit into the `link` only dependency type (which is funny given that linking is what they try to avoid :smile:)

#### Practical considerations

The first consideration I have is that we currently live with an incomplete model where, due to the lack of the `include` type dependency, we have the `link` dependency taking care also of its semantic. That said, I don't think we should encourage users to put `depends_on('libfoo', type='link')` because it seems the best fit right now for C and C++ libraries - we should instead encourage them to stick with the default in that case.

The rationale is that currently we have a default that has semantically no meaning - all dependencies should specify something else because none fits exactly `('build', 'link')`. What I think we should do is instead either:

1. Change "right now" the default dependency type to just `link` if we want to have `link` dependencies inject include paths also in the future
2. Implement the `include` dependency type and change the default to `('include', 'link')` if we want  only `include` dependency types to inject include paths

My personal preference goes to 2 for the record.

Another consideration is that `include` only dependencies should be such that they can be pruned after an install to get a minimal size installation of a given software stack. Differently from what happens for `build` type dependencies though they should still be traced and unique in the DAG.

Finally, I hope that the wall of text above makes sense to you and turns this discussion on again! :slightly_smiling_face:

**Comment by adamjstewart** [comment]
I mostly agree with @alalazo, we should add an `include` deptype. There are a lot of X11 header-only libraries (`spack list proto`) that could be uninstalled after they are used to build another package. I'm not sure whether the default should have `build` or not. One additional thing to consider is #12842. There are some test dependencies that need to be RPATHed.
